File: backtracking/combination_sum_II.40.cpp
#include <vector>

class Solution {
    public:
        std::vector<std::vector<int>> combinationSum(std::vector<int> candidates, int target)  {
            std::vector<std::vector<int>> result;
            std::sort(candidates.begin(), candidates.end());
            std::vector<int> current;
            backtrack(result, current, target, 0);
            return result;
        }
    private:
        void backtrack(std::vector<std::vector<int>>& result, std::vector<int>& current, int rest, int start,
                       std::vector<int>& candidates) {
            if (rest == 0) {
                result.push_back(current);
                return;
            }
            for (int i = start; i < candidates.size(); i++) {
                if (candidates[i] > rest) {
                    break;
                }
                if (i > start && candidates[i] == candidates[i-1]) {
                    continue;
                }
                current.push_back(candidates[i]);
                backtrack(result, current, rest - candidates[i], i + 1, candidates);
                current.pop_back();
            }
        }
}

File: backtracking/combination_sum_II.40.py

class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        result = []
        candidates.sort()
        n = len(candidates)
        def backtrack(current, rest, start):
            if rest == 0:
                result.append(current[:])
                return
            for i in range(start, n):
                if candidates[i] > rest:
                    return
                if i > start and candidates[i] == candidates[i-1]:
                    continue
                current.append(candidates[i])
                backtrack(current, rest - candidates[i], i + 1)
                current.pop()
        backtrack([], target, 0)
        return result